{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting gffpandas\n",
      "  Downloading gffpandas-1.2.0.tar.gz (178 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m178.8/178.8 kB\u001b[0m \u001b[31m12.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hBuilding wheels for collected packages: gffpandas\n",
      "  Building wheel for gffpandas (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for gffpandas: filename=gffpandas-1.2.0-py2.py3-none-any.whl size=6248 sha256=6dd3ad884f41e21f24f899e25e8ae44b9f4d210d0530de4e27a1c5f62362dd92\n",
      "  Stored in directory: /mnt/nfs01/home/j_wang@intra.igr.fr/.cache/pip/wheels/57/87/f1/1d0c74fbc5151562ba7953dc110a7d8c63c6c3229d025bc8cd\n",
      "Successfully built gffpandas\n",
      "Installing collected packages: gffpandas\n",
      "Successfully installed gffpandas-1.2.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install gffpandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gffpandas.gffpandas as gffpd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## GRCh37, gencode, Release 44 \n",
    "## Comprehensive gene annotation\n",
    "## https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/GRCh37_mapping/gencode.v44lift37.annotation.gtf.gz\n",
    "## https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/GRCh37_mapping/gencode.v44lift37.annotation.gff3.gz\n",
    "gencode_df = gffpd.read_gff3(\"resources/gencode.v44lift37.annotation.gff3\").df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "triplets = [\"AAA\", \"AAC\", \"AAG\", \"AAT\", \"CAA\", \"CAC\", \"CAG\", \"CAT\",\n",
    "            \"GAA\", \"GAC\", \"GAG\", \"GAT\", \"TAA\", \"TAC\", \"TAG\", \"TAT\", \n",
    "            \"ACA\", \"ACC\", \"ACG\", \"ACT\", \"CCA\", \"CCC\", \"CCG\", \"CCT\", \n",
    "            \"GCA\", \"GCC\", \"GCG\", \"GCT\", \"TCA\", \"TCC\", \"TCG\", \"TCT\", \n",
    "            \"AGA\", \"AGC\", \"AGG\", \"AGT\", \"CGA\", \"CGC\", \"CGG\", \"CGT\", \n",
    "            \"GGA\", \"GGC\", \"GGG\", \"GGT\", \"TGA\", \"TGC\", \"TGG\", \"TGT\", \n",
    "            \"ATA\", \"ATC\", \"ATG\", \"ATT\", \"CTA\", \"CTC\", \"CTG\", \"CTT\", \n",
    "            \"GTA\", \"GTC\", \"GTG\", \"GTT\", \"TTA\", \"TTC\", \"TTG\", \"TTT\"]\n",
    "\n",
    "triplets_user = [\"AAA\", \"AAC\", \"AAG\", \"AAT\", \"ACA\", \"ACC\", \"ACG\", \"ACT\", \n",
    "                 \"AGA\", \"AGC\", \"AGG\", \"AGT\", \"ATA\", \"ATC\", \"ATG\", \"ATT\", \n",
    "                 \"CAA\", \"CAC\", \"CAG\", \"CAT\", \"CCA\", \"CCC\", \"CCG\", \"CCT\", \n",
    "                 \"CGA\", \"CGC\", \"CGG\", \"CGT\", \"CTA\", \"CTC\", \"CTG\", \"CTT\", \n",
    "                 \"GAA\", \"GAC\", \"GAG\", \"GAT\", \"GCA\", \"GCC\", \"GCG\", \"GCT\", \n",
    "                 \"GGA\", \"GGC\", \"GGG\", \"GGT\", \"GTA\", \"GTC\", \"GTG\", \"GTT\", \n",
    "                 \"TAA\", \"TAC\", \"TAG\", \"TAT\", \"TCA\", \"TCC\", \"TCG\", \"TCT\", \n",
    "                 \"TGA\", \"TGC\", \"TGG\", \"TGT\", \"TTA\", \"TTC\", \"TTG\", \"TTT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gencode_attr_value(x, key):\n",
    "    for pair in x.split(\";\"):\n",
    "        (k,v) = pair.split(\"=\")\n",
    "        if k == key:\n",
    "            return v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_df = gencode_df[gencode_df.type == \"exon\"]\n",
    "\n",
    "gencode_df[\"gene_name\"]    = gencode_df.attributes.apply(lambda x: get_gencode_attr_value(x, \"gene_name\"))\n",
    "gencode_df[\"exon_number\"]  = gencode_df.attributes.apply(lambda x: get_gencode_attr_value(x, \"exon_number\"))\n",
    "gencode_df[\"exon_id\"]      = gencode_df.attributes.apply(lambda x: get_gencode_attr_value(x, \"ID\"))\n",
    "gencode_df[\"transcript_id\"]= gencode_df[\"exon_id\"].apply(lambda x: x.split(\":\")[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### build bed file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_bed_dataframe(fd, bed_df):\n",
    "    #### ATTENTION !!!\n",
    "    #### convert 1-based to 0-based \n",
    "    #### start - 1\n",
    "    for ind,row in bed_df.iterrows():\n",
    "        fd.write(\"%s\\t%d\\t%d\\t%s\\t%s\\t%s\\n\"%(row.Chromosome, row.chromStart_ext - 1, row.chromEnd_ext, row.gene_name, row.score, row.strand))\n",
    "\n",
    "def ext_chromStart_splice_sites(x):\n",
    "    if x.exon_number == 1:\n",
    "        # extend only context of first codon\n",
    "        return x.chromStart - 1\n",
    "    else:\n",
    "        # extend the start to splice sites if it is NOT the first codon\n",
    "        return x.chromStart - 1 - 2\n",
    "    \n",
    "def ext_chromEnd_splice_sites(x):\n",
    "    if x.exon_number == x.transcript_size:\n",
    "        return x.chromEnd + 1\n",
    "    else:\n",
    "        # extend the end to splice sites if it is NOT the last codon\n",
    "        return x.chromEnd + 1 + 2\n",
    "\n",
    "def assemble_into_bed_format(chm:str, start:int, start_ext:int, end:int, end_ext:int, group_size:int, \n",
    "                             index:int, name:str, score:str, strand:str):\n",
    "    ## bed format: \n",
    "    ## Chromosome: (e.g. chr3, chrY, chr2_random) or scaffold (e.g. scaffold10671) name\n",
    "    ## chromStart: Start coordinate on the chromosome or scaffold for the sequence considered (the first base on the chromosome is numbered 0)\n",
    "    ## chromEnd:   End coordinate on the chromosome or scaffold for the sequence considered. This position is non-inclusive, unlike chromStart.\n",
    "    ## group_size: total number of exons of the transcription\n",
    "    ## index: index number of exons in the transcription\n",
    "    ## name: Name of the line in the BED file\n",
    "    ## score: Score between 0 and 1000\n",
    "    ## strand: DNA strand orientation (positive [\"+\"] or negative [\"-\"] or \".\" if no strand)\n",
    "    return [chm.replace(\"chr\",\"\"), start, start_ext, end, end_ext, group_size, index, name, score, strand]\n",
    "\n",
    "def assemble_target_region_by(gene_df):\n",
    "    # chose the group which has max number of exons\n",
    "    exon_trans_id  = gene_df.groupby(\"transcript_id\").size().idxmax()\n",
    "    transcript_size= gene_df.groupby(\"transcript_id\").size().max()\n",
    "    gene_exons_df  = gene_df[gene_df.transcript_id == exon_trans_id]\n",
    "    \n",
    "    # name: \"gene_name:exons group size:exon number\"\n",
    "    bed_df = pd.DataFrame([assemble_into_bed_format(row.seq_id,row.start,row.start,row.end,row.end,transcript_size,int(row.exon_number),\n",
    "                            \"%s:%d:%d\"%(row.gene_name,transcript_size,int(row.exon_number)),row.score,row.strand) \n",
    "                                for ind,row in gene_exons_df.iterrows()],\n",
    "                          columns = [\"Chromosome\", \"chromStart\", \"chromStart_ext\", \"chromEnd\", \"chromEnd_ext\", \n",
    "                                        \"transcript_size\", \"exon_number\", \"gene_name\", \"score\", \"strand\"])\n",
    "    \n",
    "    bed_df[\"chromStart_ext\"] = bed_df.apply(ext_chromStart_splice_sites, axis = 1)\n",
    "    bed_df[\"chromEnd_ext\"]   = bed_df.apply(ext_chromEnd_splice_sites,   axis = 1)\n",
    "    \n",
    "    return bed_df \n",
    "\n",
    "\n",
    "with open(\"genes.bed\", \"w\") as fd:\n",
    "    cbase_gene_df = pd.read_table(\"resources/gene_list.txt\", header = None, names = [\"xxx\", \"gene\"])\n",
    "    for gene_name in cbase_gene_df.gene:\n",
    "        gene_df = gencode_df[gencode_df.gene_name == gene_name]\n",
    "        ## Naive: taking only identified targets\n",
    "        if len(gene_df) > 0:\n",
    "            bed_df = assemble_target_region_by(gene_df)\n",
    "            write_bed_dataframe(fd, bed_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: the index file is older than the FASTA file.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "os.system('bedtools getfasta -fi /run/user/1000/gvfs/sftp:host=flamingo.intra.igr.fr,user=j_wang/mnt/beegfs/userdata/a_ivashkin/references/genome_data/gatk/human_g1k_v37.fasta -bed genes.bed -name -tab -fo gencode_codons.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "codons_df = pd.read_csv(\"gencode_codons.tsv\", header=None, names=[\"gene_code\",\"exons_seq\"], sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "codons_df[\"gene_name\"]       = codons_df.gene_code.apply(lambda x: x.split(\":\")[0])\n",
    "codons_df[\"exon_group_size\"] = codons_df.gene_code.apply(lambda x: x.split(\":\")[1])\n",
    "codons_df[\"exon_group_id\"]   = codons_df.gene_code.apply(lambda x: x.split(\":\")[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_code</th>\n",
       "      <th>exons_seq</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_group_size</th>\n",
       "      <th>exon_group_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A3GALT2:5:1::1:33786675-33786702</td>\n",
       "      <td>CCTGAGTCCCTCCTTGAGAGCCATATG</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A3GALT2:5:2::1:33778404-33778494</td>\n",
       "      <td>TACCTGAATTTAGGGAGGCCATACAGAAACAGGCCTAAGAGGCCAA...</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A3GALT2:5:3::1:33778098-33778194</td>\n",
       "      <td>TACCAGGGACGCAGGGCACCTGTGAAGTTGTCTCTCAGCTGGGACA...</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A3GALT2:5:4::1:33777649-33777793</td>\n",
       "      <td>TACCTGCCTACAGCAAAGATAGTCAGCCCAATGGTGAGGTTCTGCT...</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A3GALT2:5:5::1:33772363-33773055</td>\n",
       "      <td>GCGCTAGTTCCGCAGCAGCCGGTACCCCTTGGGCGCCCACAGCAGT...</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          gene_code   \n",
       "0  A3GALT2:5:1::1:33786675-33786702  \\\n",
       "1  A3GALT2:5:2::1:33778404-33778494   \n",
       "2  A3GALT2:5:3::1:33778098-33778194   \n",
       "3  A3GALT2:5:4::1:33777649-33777793   \n",
       "4  A3GALT2:5:5::1:33772363-33773055   \n",
       "\n",
       "                                           exons_seq gene_name   \n",
       "0                        CCTGAGTCCCTCCTTGAGAGCCATATG   A3GALT2  \\\n",
       "1  TACCTGAATTTAGGGAGGCCATACAGAAACAGGCCTAAGAGGCCAA...   A3GALT2   \n",
       "2  TACCAGGGACGCAGGGCACCTGTGAAGTTGTCTCTCAGCTGGGACA...   A3GALT2   \n",
       "3  TACCTGCCTACAGCAAAGATAGTCAGCCCAATGGTGAGGTTCTGCT...   A3GALT2   \n",
       "4  GCGCTAGTTCCGCAGCAGCCGGTACCCCTTGGGCGCCCACAGCAGT...   A3GALT2   \n",
       "\n",
       "  exon_group_size exon_group_id  \n",
       "0               5             1  \n",
       "1               5             2  \n",
       "2               5             3  \n",
       "3               5             4  \n",
       "4               5             5  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codons_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene name: IL9R, exon_group_size: 10, exons id: 1, sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "[WARNING]: 'NNN' is not in list\n",
      "gene name: IL9R, exon_group_size: 10, exons id: 10, sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "[WARNING]: 'NNN' is not in list\n",
      "gene name: IL9R, exon_group_size: 10, exons id: 2, sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "[WARNING]: 'NNN' is not in list\n",
      "gene name: IL9R, exon_group_size: 10, exons id: 3, sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "[WARNING]: 'NNN' is not in list\n",
      "gene name: IL9R, exon_group_size: 10, exons id: 4, sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "[WARNING]: 'NNN' is not in list\n",
      "gene name: IL9R, exon_group_size: 10, exons id: 5, sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "[WARNING]: 'NNN' is not in list\n",
      "gene name: IL9R, exon_group_size: 10, exons id: 6, sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "[WARNING]: 'NNN' is not in list\n",
      "gene name: IL9R, exon_group_size: 10, exons id: 7, sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "[WARNING]: 'NNN' is not in list\n",
      "gene name: IL9R, exon_group_size: 10, exons id: 8, sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "[WARNING]: 'NNN' is not in list\n",
      "gene name: IL9R, exon_group_size: 10, exons id: 9, sequence: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "[WARNING]: 'NNN' is not in list\n"
     ]
    }
   ],
   "source": [
    "## problem: splice sites may be located in a codon. \n",
    "## solution: store splice sites and exons seperately \n",
    "\n",
    "def get_splice_sites_context(gene_group): \n",
    "    \n",
    "    spliceites = []\n",
    "    \n",
    "    for ind, row in gene_group.sort_values(['exon_group_id'],ascending=True).iterrows(): \n",
    "        try: \n",
    "            if int(row.exon_group_size) == 1:\n",
    "                # if there is only one exon,\n",
    "                # then do nothing\n",
    "                continue\n",
    "                \n",
    "            elif int(row.exon_group_id) == 1:\n",
    "                # splice site of first exon \n",
    "                spliceite_context_1 = triplets.index(row.exons_seq[-4:-1])\n",
    "                splice_site_context_2 = triplets.index(row.exons_seq[-3:])\n",
    "                spliceites.append([spliceite_context_1, spliceite_context_2])\n",
    "            \n",
    "            elif int(row.exon_group_id) == int(row.exon_group_size):\n",
    "                # splicee of last exon \n",
    "                spliceite_context_1 = triplets.index(row.exons_seq[0:3])\n",
    "                splice_site_context_2 = triplets.index(row.exons_seq[1:4])\n",
    "                spliceites.append([spliceite_context_1, spliceite_context_2])\n",
    "\n",
    "            else:\n",
    "                # splice site on the left of exon\n",
    "                spliceite_context_1 = triplets.index(row.exons_seq[0:3])\n",
    "                spliceite_context_2 = triplets.index(row.exons_seq[1:4])\n",
    "                splice_sites.append([splice_site_context_1, splice_site_context_2])\n",
    "\n",
    "                # splice site on the right of exon\n",
    "                splice_site_context_1 = triplets.index(row.exons_seq[-4:-1])\n",
    "                splice_site_context_2 = triplets.index(row.exons_seq[-3:])\n",
    "                splice_sites.append([splice_site_context_1, splice_site_context_2])\n",
    "                \n",
    "        except ValueError as err:\n",
    "            print(\"gene name: %s, exon_group_size: %s, exons id: %s, sequence: %s\"%(row.gene_name, row.exon_group_size, row.exon_group_id, row.exons_seq))\n",
    "            print(\"[WARNING]:\", err)\n",
    "            \n",
    "    return splice_sites\n",
    "    \n",
    "with open(\"splice_sites_by_gene.txt\", \"w\") as fd:\n",
    "    \n",
    "    for gene_name, gene_group in codons_df.groupby(\"gene_name\"):\n",
    "        splice_sites = get_splice_sites_context(gene_group)            \n",
    "        \n",
    "        fd.write(\"gene\\t%s\\n\"%gene_name)\n",
    "        \n",
    "        for ss in splice_sites:\n",
    "            fd.write(\"%d\\t%d\\n\"%(ss[0],ss[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68478</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>33786677</td>\n",
       "      <td>33786699</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000442999.3:1;Parent=ENST0000044...</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>1</td>\n",
       "      <td>exon:ENST00000442999.3:1</td>\n",
       "      <td>ENST00000442999.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68481</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>33778408</td>\n",
       "      <td>33778491</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000442999.3:2;Parent=ENST0000044...</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>2</td>\n",
       "      <td>exon:ENST00000442999.3:2</td>\n",
       "      <td>ENST00000442999.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68483</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>33778102</td>\n",
       "      <td>33778191</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000442999.3:3;Parent=ENST0000044...</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>3</td>\n",
       "      <td>exon:ENST00000442999.3:3</td>\n",
       "      <td>ENST00000442999.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68485</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>33777653</td>\n",
       "      <td>33777790</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000442999.3:4;Parent=ENST0000044...</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>4</td>\n",
       "      <td>exon:ENST00000442999.3:4</td>\n",
       "      <td>ENST00000442999.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68487</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>33772367</td>\n",
       "      <td>33773054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000442999.3:5;Parent=ENST0000044...</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>5</td>\n",
       "      <td>exon:ENST00000442999.3:5</td>\n",
       "      <td>ENST00000442999.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seq_id  source  type     start       end score strand phase   \n",
       "68478   chr1  HAVANA  exon  33786677  33786699     .      -     .  \\\n",
       "68481   chr1  HAVANA  exon  33778408  33778491     .      -     .   \n",
       "68483   chr1  HAVANA  exon  33778102  33778191     .      -     .   \n",
       "68485   chr1  HAVANA  exon  33777653  33777790     .      -     .   \n",
       "68487   chr1  HAVANA  exon  33772367  33773054     .      -     .   \n",
       "\n",
       "                                              attributes gene_name   \n",
       "68478  ID=exon:ENST00000442999.3:1;Parent=ENST0000044...   A3GALT2  \\\n",
       "68481  ID=exon:ENST00000442999.3:2;Parent=ENST0000044...   A3GALT2   \n",
       "68483  ID=exon:ENST00000442999.3:3;Parent=ENST0000044...   A3GALT2   \n",
       "68485  ID=exon:ENST00000442999.3:4;Parent=ENST0000044...   A3GALT2   \n",
       "68487  ID=exon:ENST00000442999.3:5;Parent=ENST0000044...   A3GALT2   \n",
       "\n",
       "      exon_number                   exon_id      transcript_id  \n",
       "68478           1  exon:ENST00000442999.3:1  ENST00000442999.3  \n",
       "68481           2  exon:ENST00000442999.3:2  ENST00000442999.3  \n",
       "68483           3  exon:ENST00000442999.3:3  ENST00000442999.3  \n",
       "68485           4  exon:ENST00000442999.3:4  ENST00000442999.3  \n",
       "68487           5  exon:ENST00000442999.3:5  ENST00000442999.3  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_df[gencode_df.gene_name==\"A3GALT2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
