{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "import BCBio\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gffpandas.gffpandas as gffpd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.6;gene_id=ENSG00000223972.6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000450305.2;Parent=ENSG00000223972.6;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000450305.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000450305.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12697</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000450305.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seq_id  source        type  start    end score strand phase   \n",
       "0   chr1  HAVANA        gene  12010  13670     .      +     .  \\\n",
       "1   chr1  HAVANA  transcript  12010  13670     .      +     .   \n",
       "2   chr1  HAVANA        exon  12010  12057     .      +     .   \n",
       "3   chr1  HAVANA        exon  12179  12227     .      +     .   \n",
       "4   chr1  HAVANA        exon  12613  12697     .      +     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=ENSG00000223972.6;gene_id=ENSG00000223972.6...  \n",
       "1  ID=ENST00000450305.2;Parent=ENSG00000223972.6;...  \n",
       "2  ID=exon:ENST00000450305.2:1;Parent=ENST0000045...  \n",
       "3  ID=exon:ENST00000450305.2:2;Parent=ENST0000045...  \n",
       "4  ID=exon:ENST00000450305.2:3;Parent=ENST0000045...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## GRCh37, gencode, Release 44 \n",
    "## Comprehensive gene annotation\n",
    "## https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/GRCh37_mapping/gencode.v44lift37.annotation.gtf.gz\n",
    "## https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/GRCh37_mapping/gencode.v44lift37.annotation.gff3.gz\n",
    "\n",
    "gencode_gffdf = gffpd.read_gff3(\"gencode.v44lift37.annotation.gff3\").df\n",
    "gencode_gffdf.head()\n",
    "\n",
    "## GRCh37, ensembl human, release-110\n",
    "## https://ftp.ensembl.org/pub/release-110/gff3/homo_sapiens/\n",
    "## https://ftp.ensembl.org/pub/release-110/gtf/homo_sapiens/\n",
    "\n",
    "# with open(\"Homo_sapiens.GRCh37.87.gtf\") as fd:\n",
    "#     for rec in BCBio.GFF.parse(fd):\n",
    "#         print(rec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## HGNC: HUGO Gene Nomenclature Committee\n",
    "## https://www.genenames.org/download/archive/#!/#tocAnchor-1-2\n",
    "## https://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/tsv/hgnc_complete_set.txt\n",
    "hgnc_df = pd.read_csv(\"hgnc_complete_set.txt\", sep = \"\\t\", header=0, keep_default_na=False)\n",
    "\n",
    "# gene_alias_df[[\"hgnc_id\", \"symbol\",\"name\",\"alias_symbol\",\"alias_name\",\"prev_symbol\",\"prev_name\"]][gene_alias_df.hgnc_id == \"HGNC:33584\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>alias_name</th>\n",
       "      <th>prev_symbol</th>\n",
       "      <th>prev_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td></td>\n",
       "      <td>NCRNA00181|A1BGAS|A1BG-AS</td>\n",
       "      <td>non-protein coding RNA 181|A1BG antisense RNA ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>ACF|ASP|ACF64|ACF65|APOBEC1CF</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>FWP007|S863-7|CPAMD5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HGNC:27057</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>A2M antisense RNA 1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>A2M antisense RNA 1 (non-protein coding)|A2M a...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      hgnc_id    symbol                            name   \n",
       "0      HGNC:5      A1BG          alpha-1-B glycoprotein  \\\n",
       "1  HGNC:37133  A1BG-AS1            A1BG antisense RNA 1   \n",
       "2  HGNC:24086      A1CF  APOBEC1 complementation factor   \n",
       "3      HGNC:7       A2M           alpha-2-macroglobulin   \n",
       "4  HGNC:27057   A2M-AS1             A2M antisense RNA 1   \n",
       "\n",
       "                    alias_symbol alias_name                prev_symbol   \n",
       "0                                                                       \\\n",
       "1                       FLJ23569             NCRNA00181|A1BGAS|A1BG-AS   \n",
       "2  ACF|ASP|ACF64|ACF65|APOBEC1CF                                         \n",
       "3           FWP007|S863-7|CPAMD5                                         \n",
       "4                                                                        \n",
       "\n",
       "                                           prev_name  \n",
       "0                                                     \n",
       "1  non-protein coding RNA 181|A1BG antisense RNA ...  \n",
       "2                                                     \n",
       "3                                                     \n",
       "4  A2M antisense RNA 1 (non-protein coding)|A2M a...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_alias_df = hgnc_df[[\"hgnc_id\", \"symbol\", \"name\", \"alias_symbol\", \"alias_name\", \"prev_symbol\", \"prev_name\"]]\n",
    "gene_alias_df = gene_alias_df.astype(str)\n",
    "gene_alias_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>xxx</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene</td>\n",
       "      <td>A3GALT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene</td>\n",
       "      <td>AADACL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene</td>\n",
       "      <td>AADACL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene</td>\n",
       "      <td>ABCA4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene</td>\n",
       "      <td>ABCB10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    xxx     gene\n",
       "0  gene  A3GALT2\n",
       "1  gene  AADACL3\n",
       "2  gene  AADACL4\n",
       "3  gene    ABCA4\n",
       "4  gene   ABCB10"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list = pd.read_table(\"gene_list.txt\", header=None, names=[\"xxx\", \"gene\"])\n",
    "gene_list.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_gffdf = gencode_gffdf[gencode_gffdf.type == \"exon\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_value(x, key):\n",
    "    for pair in x.split(\";\"):\n",
    "        (k,v) = pair.split(\"=\")\n",
    "        if k == key:\n",
    "            return v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_group_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000450305.2:1;Parent=ENST0000045...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>exon:ENST00000450305.2:1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000450305.2:2;Parent=ENST0000045...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>exon:ENST00000450305.2:2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12697</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000450305.2:3;Parent=ENST0000045...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>3</td>\n",
       "      <td>exon:ENST00000450305.2:3</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12975</td>\n",
       "      <td>13052</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000450305.2:4;Parent=ENST0000045...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>4</td>\n",
       "      <td>exon:ENST00000450305.2:4</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>13374</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000450305.2:5;Parent=ENST0000045...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>5</td>\n",
       "      <td>exon:ENST00000450305.2:5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seq_id  source  type  start    end score strand phase   \n",
       "2   chr1  HAVANA  exon  12010  12057     .      +     .  \\\n",
       "3   chr1  HAVANA  exon  12179  12227     .      +     .   \n",
       "4   chr1  HAVANA  exon  12613  12697     .      +     .   \n",
       "5   chr1  HAVANA  exon  12975  13052     .      +     .   \n",
       "6   chr1  HAVANA  exon  13221  13374     .      +     .   \n",
       "\n",
       "                                          attributes gene_name exon_number   \n",
       "2  ID=exon:ENST00000450305.2:1;Parent=ENST0000045...   DDX11L1           1  \\\n",
       "3  ID=exon:ENST00000450305.2:2;Parent=ENST0000045...   DDX11L1           2   \n",
       "4  ID=exon:ENST00000450305.2:3;Parent=ENST0000045...   DDX11L1           3   \n",
       "5  ID=exon:ENST00000450305.2:4;Parent=ENST0000045...   DDX11L1           4   \n",
       "6  ID=exon:ENST00000450305.2:5;Parent=ENST0000045...   DDX11L1           5   \n",
       "\n",
       "                    exon_id      exon_group_id  \n",
       "2  exon:ENST00000450305.2:1  ENST00000450305.2  \n",
       "3  exon:ENST00000450305.2:2  ENST00000450305.2  \n",
       "4  exon:ENST00000450305.2:3  ENST00000450305.2  \n",
       "5  exon:ENST00000450305.2:4  ENST00000450305.2  \n",
       "6  exon:ENST00000450305.2:5  ENST00000450305.2  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_gffdf[\"gene_name\"]  = gencode_gffdf.attributes.apply(lambda x: get_value(x, \"gene_name\"))\n",
    "gencode_gffdf[\"exon_number\"]= gencode_gffdf.attributes.apply(lambda x: get_value(x, \"exon_number\"))\n",
    "gencode_gffdf[\"exon_id\"]    = gencode_gffdf.attributes.apply(lambda x: get_value(x, \"ID\"))\n",
    "gencode_gffdf[\"exon_group_id\"] = gencode_gffdf[\"exon_id\"].apply(lambda x: x.split(\":\")[1])\n",
    "gencode_gffdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "\n",
    "# [\"hgnc_id\", \"symbol\",\"name\",\"alias_symbol\",\"alias_name\",\"prev_symbol\",\"prev_name\"]\n",
    "\n",
    "def is_symbol(gene):\n",
    "    rec = gene_alias_df[gene_alias_df.symbol == gene]\n",
    "    if len(rec) > 0 :\n",
    "        return True\n",
    "    return False\n",
    "\n",
    "def convert2symbol(gene):\n",
    "    for symbol in [\"alias_symbol\",\"prev_symbol\",\"name\",\"alias_name\",\"prev_name\"]:\n",
    "        rec = gene_alias_df[gene_alias_df[symbol].apply(lambda x: True if x.split(\"|\").count(gene) > 0 else False)]\n",
    "        if len(rec) > 0 :\n",
    "            return rec.symbol.iloc[0]\n",
    "            \n",
    "    warnings.warn(\"Symbol %s is not matched in HGNC!\"%gene, ResourceWarning)\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## correct gencode gene name to standard symbol\n",
    "# symbol_convert_dict = {}\n",
    "# for gene_name in set(gencode_gffdf.gene_name):\n",
    "#     if not is_symbol(gene_name):\n",
    "#         symbol_convert_dict[gene_name] = convert2symbol(gene_name)      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(symbol_convert_dict)\n",
    "# gencode_gffdf[gencode_gffdf.gene_name == \"ENSG00000283839\"] #.attributes.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assemble_into_bed_format(chm, start, end, name, score, strand):\n",
    "    ## bed format: \n",
    "    ## Chromosome: (e.g. chr3, chrY, chr2_random) or scaffold (e.g. scaffold10671) name\n",
    "    ## chromStart: Start coordinate on the chromosome or scaffold for the sequence considered (the first base on the chromosome is numbered 0)\n",
    "    ## chromEnd:   End coordinate on the chromosome or scaffold for the sequence considered. This position is non-inclusive, unlike chromStart.\n",
    "    ## name: Name of the line in the BED file\n",
    "    ## score: Score between 0 and 1000\n",
    "    ## strand: DNA strand orientation (positive [\"+\"] or negative [\"-\"] or \".\" if no strand)\n",
    "    return [chm.replace(\"chr\",\"\"), int(start), int(end), name, score, strand]\n",
    "\n",
    "def assemble_target_region_by(gene_name):\n",
    "    gene_exons_df = gencode_gffdf[gencode_gffdf.gene_name == gene_name]\n",
    "    if len(gene_exons_df) > 0 :\n",
    "        # chose the group which has max number of exons\n",
    "        exon_group_id  = gene_exons_df.groupby(\"exon_group_id\").size().idxmax()\n",
    "        exon_group_size= gene_exons_df.groupby(\"exon_group_id\").size().max()\n",
    "        gene_exons_df  = gene_exons_df[gene_exons_df.exon_group_id == exon_group_id]\n",
    "        \n",
    "        # name: \"gene_name:exons group size:exon number\"\n",
    "        return pd.DataFrame([ assemble_into_bed_format(row.seq_id, row.start, row.end, \n",
    "                                        \"%s:%d:%d\"%(row.gene_name,exon_group_size,int(row.exon_number)),\n",
    "                                        row.score, row.strand) for ind,row in gene_exons_df.iterrows()], \n",
    "                               columns = [\"Chromosome\", \"chromStart\", \"chromEnd\", \"gene_name\", \"score\", \"strand\"])\n",
    "        \n",
    "    else:\n",
    "        return pd.DataFrame([], columns = [\"Chromosome\", \"chromStart\", \"chromEnd\", \"gene_name\", \"score\", \"strand\"])\n",
    "    \n",
    "def write_bed_dataframe(fd, bed_df):\n",
    "    for ind,row in bed_df.iterrows():\n",
    "        fd.write(\"%s\\t%d\\t%d\\t%s\\t%s\\t%s\\n\"%(row.Chromosome, row.chromStart, row.chromEnd, row.gene_name, row.score, row.strand))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "\n",
    "gene_missmatch_list = []\n",
    "\n",
    "with open(\"genes.bed\", \"w\") as fd:\n",
    "    for gene_name in gene_list.gene:\n",
    "    \n",
    "        bed_df = assemble_target_region_by(gene_name)\n",
    "    \n",
    "        if len(bed_df) > 0:\n",
    "            write_bed_dataframe(fd, bed_df)\n",
    "            continue\n",
    "            \n",
    "        if not is_symbol(gene_name): \n",
    "            # convert gene name to stardard symbol\n",
    "            symbol = convert2symbol(gene_name)\n",
    "            bed_df   = assemble_target_region_by(symbol)\n",
    "            if len(bed_df) > 0:\n",
    "                write_bed_dataframe(fd, bed_df)\n",
    "                continue\n",
    "        \n",
    "        # gene_name in genocode is not standard symbol\n",
    "        # TODO: match all the possible symbols\n",
    "        # TODO: raise warning and correct HGNC(HUGO Gene Nomenclature Committee)\n",
    "        # warnings.warn(\"gene name [%s] is mismatched in HGNC!\\n\"%gene_name)\n",
    "        gene_missmatch_list.append(gene_name)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "mismatch_df = pd.DataFrame(gene_missmatch_list, columns=[\"mismatch\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "mismatch_df[\"is_symbol\"] = mismatch_df.mismatch.apply(is_symbol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mismatch</th>\n",
       "      <th>is_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AK022898</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AK095633</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AK097814</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AK125437</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AK127270</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604</th>\n",
       "      <td>LOC100129520</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>605</th>\n",
       "      <td>LOC100132831</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>606</th>\n",
       "      <td>LOC100133957</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>U00684</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>608</th>\n",
       "      <td>CD24</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>609 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         mismatch  is_symbol\n",
       "0        AK022898      False\n",
       "1        AK095633      False\n",
       "2        AK097814      False\n",
       "3        AK125437      False\n",
       "4        AK127270      False\n",
       "..            ...        ...\n",
       "604  LOC100129520      False\n",
       "605  LOC100132831      False\n",
       "606  LOC100133957      False\n",
       "607        U00684      False\n",
       "608          CD24       True\n",
       "\n",
       "[609 rows x 2 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mismatch_df # [mismatch_df.mismatch.apply(lambda x: x[:2] == \"AK\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_group_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3328939</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47518232</td>\n",
       "      <td>47518579</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000591832.1:1;Parent=ENST0000059...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>1</td>\n",
       "      <td>exon:ENST00000591832.1:1</td>\n",
       "      <td>ENST00000591832.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328940</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47519294</td>\n",
       "      <td>47519510</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000591832.1:2;Parent=ENST0000059...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>2</td>\n",
       "      <td>exon:ENST00000591832.1:2</td>\n",
       "      <td>ENST00000591832.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328942</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47518252</td>\n",
       "      <td>47519776</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000590504.1:1;Parent=ENST0000059...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>1</td>\n",
       "      <td>exon:ENST00000590504.1:1</td>\n",
       "      <td>ENST00000590504.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328944</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47518490</td>\n",
       "      <td>47518579</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000664698.1:1;Parent=ENST0000066...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>1</td>\n",
       "      <td>exon:ENST00000664698.1:1</td>\n",
       "      <td>ENST00000664698.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328945</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47531903</td>\n",
       "      <td>47531992</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000664698.1:2;Parent=ENST0000066...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>2</td>\n",
       "      <td>exon:ENST00000664698.1:2</td>\n",
       "      <td>ENST00000664698.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328946</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47548461</td>\n",
       "      <td>47551725</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000664698.1:3;Parent=ENST0000066...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>3</td>\n",
       "      <td>exon:ENST00000664698.1:3</td>\n",
       "      <td>ENST00000664698.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328948</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47518634</td>\n",
       "      <td>47518800</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000670339.1:1;Parent=ENST0000067...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>1</td>\n",
       "      <td>exon:ENST00000670339.1:1</td>\n",
       "      <td>ENST00000670339.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328949</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47519294</td>\n",
       "      <td>47521298</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000670339.1:2;Parent=ENST0000067...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>2</td>\n",
       "      <td>exon:ENST00000670339.1:2</td>\n",
       "      <td>ENST00000670339.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328951</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47531939</td>\n",
       "      <td>47531992</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000658400.1:1;Parent=ENST0000065...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>1</td>\n",
       "      <td>exon:ENST00000658400.1:1</td>\n",
       "      <td>ENST00000658400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328952</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47549074</td>\n",
       "      <td>47551705</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000658400.1:2;Parent=ENST0000065...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>2</td>\n",
       "      <td>exon:ENST00000658400.1:2</td>\n",
       "      <td>ENST00000658400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328954</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47531973</td>\n",
       "      <td>47531992</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000669654.1:1;Parent=ENST0000066...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>1</td>\n",
       "      <td>exon:ENST00000669654.1:1</td>\n",
       "      <td>ENST00000669654.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328955</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47544715</td>\n",
       "      <td>47544860</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000669654.1:2;Parent=ENST0000066...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>2</td>\n",
       "      <td>exon:ENST00000669654.1:2</td>\n",
       "      <td>ENST00000669654.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328956</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>47549074</td>\n",
       "      <td>47550361</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000669654.1:3;Parent=ENST0000066...</td>\n",
       "      <td>UXT-AS1</td>\n",
       "      <td>3</td>\n",
       "      <td>exon:ENST00000669654.1:3</td>\n",
       "      <td>ENST00000669654.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seq_id  source  type     start       end score strand phase   \n",
       "3328939   chrX  HAVANA  exon  47518232  47518579     .      +     .  \\\n",
       "3328940   chrX  HAVANA  exon  47519294  47519510     .      +     .   \n",
       "3328942   chrX  HAVANA  exon  47518252  47519776     .      +     .   \n",
       "3328944   chrX  HAVANA  exon  47518490  47518579     .      +     .   \n",
       "3328945   chrX  HAVANA  exon  47531903  47531992     .      +     .   \n",
       "3328946   chrX  HAVANA  exon  47548461  47551725     .      +     .   \n",
       "3328948   chrX  HAVANA  exon  47518634  47518800     .      +     .   \n",
       "3328949   chrX  HAVANA  exon  47519294  47521298     .      +     .   \n",
       "3328951   chrX  HAVANA  exon  47531939  47531992     .      +     .   \n",
       "3328952   chrX  HAVANA  exon  47549074  47551705     .      +     .   \n",
       "3328954   chrX  HAVANA  exon  47531973  47531992     .      +     .   \n",
       "3328955   chrX  HAVANA  exon  47544715  47544860     .      +     .   \n",
       "3328956   chrX  HAVANA  exon  47549074  47550361     .      +     .   \n",
       "\n",
       "                                                attributes gene_name   \n",
       "3328939  ID=exon:ENST00000591832.1:1;Parent=ENST0000059...   UXT-AS1  \\\n",
       "3328940  ID=exon:ENST00000591832.1:2;Parent=ENST0000059...   UXT-AS1   \n",
       "3328942  ID=exon:ENST00000590504.1:1;Parent=ENST0000059...   UXT-AS1   \n",
       "3328944  ID=exon:ENST00000664698.1:1;Parent=ENST0000066...   UXT-AS1   \n",
       "3328945  ID=exon:ENST00000664698.1:2;Parent=ENST0000066...   UXT-AS1   \n",
       "3328946  ID=exon:ENST00000664698.1:3;Parent=ENST0000066...   UXT-AS1   \n",
       "3328948  ID=exon:ENST00000670339.1:1;Parent=ENST0000067...   UXT-AS1   \n",
       "3328949  ID=exon:ENST00000670339.1:2;Parent=ENST0000067...   UXT-AS1   \n",
       "3328951  ID=exon:ENST00000658400.1:1;Parent=ENST0000065...   UXT-AS1   \n",
       "3328952  ID=exon:ENST00000658400.1:2;Parent=ENST0000065...   UXT-AS1   \n",
       "3328954  ID=exon:ENST00000669654.1:1;Parent=ENST0000066...   UXT-AS1   \n",
       "3328955  ID=exon:ENST00000669654.1:2;Parent=ENST0000066...   UXT-AS1   \n",
       "3328956  ID=exon:ENST00000669654.1:3;Parent=ENST0000066...   UXT-AS1   \n",
       "\n",
       "        exon_number                   exon_id      exon_group_id  \n",
       "3328939           1  exon:ENST00000591832.1:1  ENST00000591832.1  \n",
       "3328940           2  exon:ENST00000591832.1:2  ENST00000591832.1  \n",
       "3328942           1  exon:ENST00000590504.1:1  ENST00000590504.1  \n",
       "3328944           1  exon:ENST00000664698.1:1  ENST00000664698.1  \n",
       "3328945           2  exon:ENST00000664698.1:2  ENST00000664698.1  \n",
       "3328946           3  exon:ENST00000664698.1:3  ENST00000664698.1  \n",
       "3328948           1  exon:ENST00000670339.1:1  ENST00000670339.1  \n",
       "3328949           2  exon:ENST00000670339.1:2  ENST00000670339.1  \n",
       "3328951           1  exon:ENST00000658400.1:1  ENST00000658400.1  \n",
       "3328952           2  exon:ENST00000658400.1:2  ENST00000658400.1  \n",
       "3328954           1  exon:ENST00000669654.1:1  ENST00000669654.1  \n",
       "3328955           2  exon:ENST00000669654.1:2  ENST00000669654.1  \n",
       "3328956           3  exon:ENST00000669654.1:3  ENST00000669654.1  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_gffdf[gencode_gffdf.gene_name == \"UXT-AS1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>status</th>\n",
       "      <th>location</th>\n",
       "      <th>location_sortable</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>alias_name</th>\n",
       "      <th>...</th>\n",
       "      <th>cd</th>\n",
       "      <th>lncrnadb</th>\n",
       "      <th>enzyme_id</th>\n",
       "      <th>intermediate_filament_db</th>\n",
       "      <th>rna_central_ids</th>\n",
       "      <th>lncipedia</th>\n",
       "      <th>gtrnadb</th>\n",
       "      <th>agr</th>\n",
       "      <th>mane_select</th>\n",
       "      <th>gencc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14797</th>\n",
       "      <td>HGNC:27631</td>\n",
       "      <td>LEMD1-DT</td>\n",
       "      <td>LEMD1 divergent transcript</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1q32.1</td>\n",
       "      <td>01q32.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>HGNC:27631</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          hgnc_id    symbol                        name     locus_group   \n",
       "14797  HGNC:27631  LEMD1-DT  LEMD1 divergent transcript  non-coding RNA  \\\n",
       "\n",
       "                 locus_type    status location location_sortable alias_symbol   \n",
       "14797  RNA, long non-coding  Approved   1q32.1           01q32.1               \\\n",
       "\n",
       "      alias_name  ... cd lncrnadb enzyme_id intermediate_filament_db   \n",
       "14797             ...                                                 \\\n",
       "\n",
       "      rna_central_ids lncipedia gtrnadb         agr mane_select gencc  \n",
       "14797                                    HGNC:27631                    \n",
       "\n",
       "[1 rows x 54 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgnc_df[hgnc_df.symbol == \"LEMD1-DT\"]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
