{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.optimize import minimize\n",
    "import scipy.special as sp\n",
    "import sys\n",
    "import itertools as it\n",
    "import math\n",
    "import gzip\n",
    "import numpy as np\n",
    "# import argparse\n",
    "# import os\n",
    "import random\n",
    "# import subprocess\n",
    "import mpmath as mp\n",
    "import scipy.stats as st\n",
    "import itertools as it\n",
    "# import subprocess\n",
    "# import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "\n",
    "# triplets_user = [\"AAA\", \"AAC\", \"AAG\", \"AAT\", \"ACA\", \"ACC\", \"ACG\", \"ACT\", \n",
    "#                  \"AGA\", \"AGC\", \"AGG\", \"AGT\", \"ATA\", \"ATC\", \"ATG\", \"ATT\", \n",
    "#                  \"CAA\", \"CAC\", \"CAG\", \"CAT\", \"CCA\", \"CCC\", \"CCG\", \"CCT\", \n",
    "#                  \"CGA\", \"CGC\", \"CGG\", \"CGT\", \"CTA\", \"CTC\", \"CTG\", \"CTT\", \n",
    "#                  \"GAA\", \"GAC\", \"GAG\", \"GAT\", \"GCA\", \"GCC\", \"GCG\", \"GCT\", \n",
    "#                  \"GGA\", \"GGC\", \"GGG\", \"GGT\", \"GTA\", \"GTC\", \"GTG\", \"GTT\", \n",
    "#                  \"TAA\", \"TAC\", \"TAG\", \"TAT\", \"TCA\", \"TCC\", \"TCG\", \"TCT\", \n",
    "#                  \"TGA\", \"TGC\", \"TGG\", \"TGT\", \"TTA\", \"TTC\", \"TTG\", \"TTT\"]\n",
    "\n",
    "# bcc_all_df = pd.read_csv(\"BCC_ALL_MUTATIONS.jx.tsv\", sep=\"\\t\")\n",
    "# bcc_splice_sites_df = bcc_all_df[bcc_all_df.Variant_Classification.isin([\"Splice_Site\", \"Silent\", \"Missense_Mutation\", 'Nonsense_Mutation', '3\\'UTR', '5\\'UTR'])].reindex()\n",
    "# bcc_splice_sites_df = bcc_splice_sites_df[bcc_splice_sites_df.signature_context.str.strip() != '-' ].reindex()\n",
    "# bcc_fullset_df = bcc_splice_sites_df[[\"Hugo_Symbol\",\"Variant_Classification\", \"signature_context\"]]\n",
    "# bcc_fullset_df.loc[:,\"mut_nuc\"] = bcc_fullset_df.apply(lambda x: x.signature_context.split(\">\")[-1], axis=1)\n",
    "# bcc_fullset_df.loc[:,\"context\"] = bcc_fullset_df.apply(lambda x: x.signature_context.split(\"_\")[0],  axis=1)\n",
    "# bcc_fullset_df.columns = [\"gene\", \"mut_eff\", \"signature_context\", \"mut_nuc\", \"context\"]\n",
    "# bcc_fullset_df['mut_eff'] = bcc_fullset_df.mut_eff.str.replace(\"Missense_Mutation\",\"missense\").replace(\"Nonsense_Mutation\",\"nonsense\").replace(\"Silent\",\"coding-synon\").replace(\"3\\'UTR\",\"utr-3\").replace(\"5\\'UTR\",'utr-5')\n",
    "# bcc_fullset_df.mut_eff = bcc_fullset_df.mut_eff.str.replace(\"Splice_Site\", \"splice_site\")\n",
    "# bcc_fullset_df.context = bcc_fullset_df.context.apply(lambda x: triplets_user.index(x))\n",
    "# bcc_fullset_df[[\"gene\", \"mut_eff\", \"mut_nuc\", \"context\"]].to_csv(\"BCC_fullset_ss.txt\", sep=\"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def muttype_index(cod_ID):\n",
    "\t# 0: missense, 1: nonsense, 2: synonymous\n",
    "\tif [56,57,58,59].count(cod_ID):\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [2,2,2,2]]\n",
    "\telif [20,21,22,23].count(cod_ID):\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [2,2,2,2]]\n",
    "\telif [16,17,18,19].count(cod_ID):\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [2,2,2,2]]\n",
    "\telif [24,25,26,27].count(cod_ID):\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [2,2,2,2]]\n",
    "\telif cod_ID==40:\n",
    "\t\treturn [[0,0,0,1], [0,0,0,0], [2,2,2,2]]\n",
    "\telif [41,42,43].count(cod_ID):\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [2,2,2,2]]\n",
    "\telif cod_ID==5:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [0,0,0,2]]\n",
    "\telif cod_ID==7:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [0,2,0,0]]\n",
    "\telif cod_ID==1:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [0,0,0,2]]\n",
    "\telif cod_ID==3:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [0,2,0,0]]\n",
    "\telif cod_ID==9:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [0,0,0,2]]\n",
    "\telif cod_ID==11:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [0,2,0,0]]\n",
    "\telif cod_ID==61:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [0,0,0,2]]\n",
    "\telif cod_ID==63:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [0,2,0,0]]\n",
    "\telif cod_ID==13:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [1,0,1,2]]\n",
    "\telif cod_ID==15:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [1,2,1,0]]\n",
    "\telif cod_ID==4:\n",
    "\t\treturn [[0,0,0,1], [0,0,0,0], [0,0,2,0]]\n",
    "\telif cod_ID==6:\n",
    "\t\treturn [[0,0,0,1], [0,0,0,0], [2,0,0,0]]\n",
    "\telif cod_ID==0:\n",
    "\t\treturn [[0,0,0,1], [0,0,0,0], [0,0,2,0]]\n",
    "\telif cod_ID==2:\n",
    "\t\treturn [[0,0,0,1], [0,0,0,0], [2,0,0,0]]\n",
    "\telif cod_ID==8:\n",
    "\t\treturn [[0,0,0,1], [0,0,0,0], [0,0,2,0]]\n",
    "\telif cod_ID==10:\n",
    "\t\treturn [[0,0,0,1], [0,0,0,0], [2,0,0,0]]\n",
    "\telif cod_ID==45:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [1,0,0,2]]\n",
    "\telif cod_ID==47:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [1,2,0,0]]\n",
    "\telif [52,54].count(cod_ID):\n",
    "\t\treturn [[0,0,0,2], [0,0,0,0], [2,2,2,2]]\n",
    "\telif [53,55].count(cod_ID):\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [2,2,2,2]]\n",
    "\telif cod_ID==60:\n",
    "\t\treturn [[0,2,0,0], [1,0,1,0], [0,0,2,0]]\n",
    "\telif cod_ID==62:\n",
    "\t\treturn [[0,2,0,0], [1,0,0,0], [2,0,0,0]]\n",
    "\telif cod_ID==36:\n",
    "\t\treturn [[2,0,0,1], [0,0,0,0], [2,2,2,2]]\n",
    "\telif cod_ID==38:\n",
    "\t\treturn [[2,0,0,0], [0,0,0,0], [2,2,2,2]]\n",
    "\telif [37,39].count(cod_ID):\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [2,2,2,2]]\n",
    "\telif cod_ID==32:\n",
    "\t\treturn [[0,2,0,1], [0,0,0,0], [0,0,2,0]]\n",
    "\telif cod_ID==34:\n",
    "\t\treturn [[0,2,0,0], [0,0,0,0], [2,0,0,0]]\n",
    "\telif cod_ID==28:\n",
    "\t\treturn [[0,0,0,0], [1,0,1,0], [2,2,2,2]]\n",
    "\telif cod_ID==30:\n",
    "\t\treturn [[0,0,0,0], [1,0,0,0], [2,2,2,2]]\n",
    "\telif [29,31].count(cod_ID):\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [2,2,2,2]]\n",
    "\telif cod_ID==33:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [0,0,0,2]]\n",
    "\telif cod_ID==35:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [0,2,0,0]]\n",
    "\telif cod_ID==12:\n",
    "\t\treturn [[1,1,1,1], [0,1,2,1], [0,1,2,1]]\n",
    "\telif cod_ID==14:\n",
    "\t\treturn [[1,1,1,1], [1,1,1,1], [2,1,0,1]]\n",
    "\telif cod_ID==44:\n",
    "\t\treturn [[1,1,1,1], [2,1,0,1], [1,1,1,1]]\n",
    "\telif [48,49,51].count(cod_ID):\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [2,2,0,2]]\n",
    "\telif cod_ID==50:\n",
    "\t\treturn [[0,0,0,0], [0,0,0,0], [0,0,0,0]]\n",
    "\telif cod_ID==46:\n",
    "\t\treturn [[0,0,0,0], [1,0,0,0], [1,0,0,0]]\n",
    "\telse:\n",
    "\t\tsys.stderr.write(\"Strange codon ID: %i.\\n\" %cod_ID)\n",
    "\t\t\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_special_genes(filename):\n",
    "\tfin = open(filename)\n",
    "\tlines = fin.readlines()\n",
    "\tfin.close()\n",
    "\tc_genes = []\n",
    "\tfor line in lines:\n",
    "\t\tc_genes.append(line.strip().split()[0])\n",
    "\treturn c_genes\n",
    "\n",
    "def import_context_freqs(filename):\n",
    "\tfin = open(filename)\n",
    "\tlines = fin.readlines()\n",
    "\tfin.close()\n",
    "\toccs=[]\n",
    "\tfor line in lines:\n",
    "\t\tfield = line.strip().split()\n",
    "\t\toccs.append(float(field[1]))\n",
    "\treturn occs\n",
    "\n",
    "def import_quintuplets(filename):\n",
    "\tfin = open(filename)\n",
    "\tlines = fin.readlines()\n",
    "\tfin.close()\n",
    "\treturn [line.strip().split()[0] for line in lines]\n",
    "\n",
    "def import_maf_data(filename, context_mode):\n",
    "\tfin = open(filename)\n",
    "\tlines = fin.readlines()\n",
    "\tfin.close()\n",
    "\tmut_array=[]\n",
    "\tfor line in lines[1:]:\n",
    "\t\tfield = line.strip().split(\"\\t\")\n",
    "\t\tif len(field)!=4:\n",
    "\t\t\tsys.stderr.write(\"Number of columns in maf file not as expected (=4): %i.\\n\" %len(field))\n",
    "\t\t\tsys.exit()\n",
    "\t\t# context is 0-based; triplets mapped to legacy\n",
    "\t\tif context_mode==0:\n",
    "\t\t\tmut_array.append({\"gene\": field[0], \"muttype\": field[1], \"mutbase\": field[2].upper(), \"context\": triplets.index(triplets_user[int(field[3])])})\n",
    "\t\telse:\n",
    "\t\t\tmut_array.append({\"gene\": field[0], \"muttype\": field[1], \"mutbase\": field[2].upper(), \"context\": int(field[3])})\n",
    "\treturn mut_array\n",
    "\n",
    "\n",
    "#### This function is never called\n",
    "def import_known_genes_UCSC(filename):\n",
    "\tfin = open(filename)\n",
    "\tlines = fin.readlines()\n",
    "\tfin.close()\n",
    "\tgenes=[]\n",
    "\tfor line in lines:\n",
    "\t\tfield = line.strip().split('\\t')\n",
    "\t\tif field[1][3:]=='M' or len(field[1].split('_'))>1:\n",
    "\t\t\tcontinue\n",
    "\t\tif field[1][3:]=='X':\n",
    "\t\t\tchrnr = 23\n",
    "\t\telif field[1][3:]=='Y':\n",
    "\t\t\tchrnr = 24\n",
    "\t\telse:\n",
    "\t\t\tchrnr = int(field[1][3:])\n",
    "\t\tif len(field)!=14:\n",
    "\t\t\tsys.stderr.write(\"Unexpected line format.\\n\")\n",
    "\t\t\tsys.exit()\n",
    "\t\tif field[11]!='n/a' and field[12]!='n/a':\n",
    "\t\t\tgenes.append({\"transcript\": field[0], \"gene\": field[10], \"chr\": chrnr, \"strand\": field[2], \"genebegin\": int(field[3]), \"geneend\": int(field[4]), \"cdbegin\": int(field[5]), \"cdend\": int(field[6]), \"exoncnt\": int(field[7]), \"exonbegins\": [int(el) for el in field[8][:-1].split(',')], \"exonends\": [int(el) for el in field[9][:-1].split(',')], \"pepseq\": field[12], \"ensemblID\": field[13], \"all\": field})\n",
    "\tgenes = sorted(genes, key=lambda arg: arg[\"gene\"])\n",
    "\treturn genes\n",
    "\n",
    "def import_codons_by_gene(filename):\n",
    "\tsys.stderr.write(\"Importing codons...\\n\")\n",
    "\twith gzip.open(filename, 'rt') as fin:\n",
    "\t\tlines = fin.readlines()\n",
    "\tfin.close()\n",
    "\tc_genes=[]; cur_gene=\"bla\"; cur_cods=[]\n",
    "\tfor line in lines:\n",
    "\t\tfield = line.strip().split(\"\\t\")\n",
    "\t\tif len(field)==2:\n",
    "\t\t\tc_genes.append({\"gene\": cur_gene, \"context_triplets\": cur_cods})\n",
    "\t\t\tcur_cods=[]\n",
    "\t\t\tcur_gene = field[1]\n",
    "\t\telse:\n",
    "\t\t\tcur_cods.append([int(el) for el in field])\n",
    "\tc_genes.append({\"gene\": cur_gene, \"context_triplets\": cur_cods})\n",
    "\treturn c_genes[1:]\n",
    "\n",
    "def import_splice_sites_by_gene(filename):\n",
    "\tsys.stderr.write(\"Importing splice Sites...\\n\")\n",
    "\tc_splice_sites  = {}\n",
    "\tcur_splice_sites= None\n",
    "\tcur_cods          = []\n",
    "\twith open(filename, 'rt') as fd:\n",
    "\t\tfor line in fd:\n",
    "\t\t\tfields = line.strip().split(\"\\t\")\n",
    "\t\t\tif fields[0]==\"gene\":\n",
    "\t\t\t\tif cur_splice_sites != None:\n",
    "\t\t\t\t\tc_splice_sites[cur_splice_sites] = cur_cods\n",
    "\t\t\t\tcur_cods=[]\n",
    "\t\t\t\tcur_splice_sites = fields[1]\n",
    "\t\t\telse:\n",
    "\t\t\t\tcur_cods.append([int(el) for el in fields])\n",
    "\t\tc_splice_sites[cur_splice_sites] = cur_cods\n",
    "\treturn c_splice_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_neutral_mut_matrix_pentanucs(neut_cont_array, penta_occs_array):\n",
    "\tneutral_mut_matrix = [[0. for i in range(4)] for j in range(1024)]\n",
    "\ttotcnt=0\n",
    "\tfor el in neut_cont_array:\n",
    "\t\tneutral_mut_matrix[el[0]][el[1]] += 1.\n",
    "\t\ttotcnt+=1\n",
    "\t#\tIncluding pseudocounts:\n",
    "\tfor m in range(16):\n",
    "\t\tfor k in range(4):\n",
    "\t\t\talist = [0,1,2,3]\n",
    "\t\t\talist.remove(k)\n",
    "\t\t\tuselist = alist[:]\n",
    "\t\t\tfor i in range(16):\n",
    "\t\t\t\tind1 = m*64+k*16+i\n",
    "\t\t\t\tfor j in uselist:\n",
    "\t\t\t\t\tneutral_mut_matrix[ind1][j] += 1.\n",
    "\t\t\t\t\ttotcnt+=1\n",
    "\t# Symmetrize the probabilities:\n",
    "\tprobs_array=[[0. for i in range(4)] for j in range(1024)]\n",
    "\tfor i in range(4):\n",
    "\t\tfor j in range(4):\n",
    "\t\t\tfor k in range(4):\n",
    "\t\t\t\tfor l in range(4):\n",
    "\t\t\t\t\tfor m in range(4):\n",
    "\t\t\t\t\t\tn\t\t= 256*i+64*j+16*k+4*l+m\n",
    "\t\t\t\t\t\tnprime\t= 1023-(256*m+64*l+16*k+4*j+i)\n",
    "\t\t\t\t\t\tfor h in range(4):\n",
    "\t\t\t\t\t\t\ttotalcnt\t\t\t= neutral_mut_matrix[n][h] + neutral_mut_matrix[nprime][3-h]\n",
    "\t\t\t\t\t\t\tprobs_array[n][h]\t= totalcnt/(2.*totcnt)/((penta_occs_array[n]+penta_occs_array[nprime])/2.)\n",
    "\t\t\t\t\t\t\tprobs_array[nprime][3-h]\t= probs_array[n][h]\n",
    "\treturn probs_array\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_neutral_mut_matrix_trinucs(neut_cont_array, trinuc_occs_array):\n",
    "\tneutral_mut_matrix = [[0. for i in range(4)] for j in range(64)]\n",
    "\ttotcnt=0\n",
    "\tfor el in neut_cont_array:\n",
    "\t\tneutral_mut_matrix[el[0]][el[1]] += 1.\n",
    "\t\ttotcnt+=1\n",
    "\t# Symmetrize the probabilities:\n",
    "\tprobs_array=[[0. for i in range(4)] for j in range(64)]\n",
    "\tfor i in range(4):\n",
    "\t\tfor j in range(4):\n",
    "\t\t\tfor k in range(4):\n",
    "\t\t\t\tn\t\t= 16*j+4*i+k\n",
    "\t\t\t\tnprime\t= 63-(16*j+4*k+i)\n",
    "\t\t\t\tfor m in range(4):\n",
    "\t\t\t\t\ttotalcnt\t\t\t\t\t= neutral_mut_matrix[n][m] + neutral_mut_matrix[nprime][3-m]\n",
    "\t\t\t\t\tprobs_array[n][m]\t\t\t= totalcnt/(2.*totcnt)/((trinuc_occs_array[n]+trinuc_occs_array[nprime])/2.)\n",
    "\t\t\t\t\tprobs_array[nprime][3-m]\t= probs_array[n][m]\n",
    "\treturn probs_array\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def export_expected_observed_mks_per_gene(codon_array_by_gene, mut_array, neutral_muts_array, nuc_occs_array, context_mode):\n",
    "\t\n",
    "\tmut_coding = [m for m in mut_array if [\"missense\", \"nonsense\", \"coding-synon\"].count(m[\"muttype\"])]\n",
    "\t# sys.stderr.write(\"%i (m,k,s,c) mutations.\\n\"%(len(mut_coding)))\n",
    "\n",
    "\tgene_keys=[]; muts_by_gene=[]\n",
    "\tfor k, g in it.groupby(sorted(mut_coding, key=lambda arg: arg[\"gene\"]), key=lambda arg: arg[\"gene\"]):\n",
    "\t\tgene_keys.append(k)\n",
    "\t\tmuts_by_gene.append(list(g))\n",
    "\n",
    "\t#\tDerive the neutral transition matrix from the sum over all patients.\n",
    "\tif context_mode==0:\t\t\t\t\t#\tTRINUCLEOTIDES\n",
    "\t\tneutral_matrix = make_neutral_mut_matrix_trinucs(neutral_muts_array, nuc_occs_array)\n",
    "\n",
    "\telse:\t\t\t\t\t\t\t\t#\tPENTANUCLEOTIDES\n",
    "\t\tneutral_matrix = make_neutral_mut_matrix_pentanucs(neutral_muts_array, nuc_occs_array)\n",
    "\n",
    "\tprint(\"neutral_matrix: \\n\")\n",
    "\n",
    "\tfor el in neutral_matrix:\n",
    "\t\tfor i in range(len(el)):\n",
    "\t\t\tsys.stderr.write(\"%f\\t\" %el[i])\n",
    "\t\tsys.stderr.write(\"\\n\")\n",
    "\n",
    "\tif len(neutral_matrix)==0:\n",
    "\t\tsys.stderr.write(\"Cannot construct neutral mutation matrix.\\n\")\n",
    "\t\tsys.exit()\n",
    "\n",
    "\texp_obs_per_gene=[]\n",
    "\t\n",
    "\tfor g in range(len(codon_array_by_gene)):\n",
    "\n",
    "\t\tcur_gene\t= codon_array_by_gene[g][\"gene\"]\n",
    "\t\tcodons_gene\t= codon_array_by_gene[g][\"context_triplets\"]\n",
    "\t\tgene_len\t= len(codons_gene)*3\n",
    "\n",
    "\t\txobs = [0 for i in range(4)]\n",
    "\t\ttry:\n",
    "\t\t\tgene_muts = muts_by_gene[gene_keys.index(cur_gene)]\n",
    "\t\t\t#\tSum observed mutations in categories over patients.\n",
    "\t\t\tfor mut in gene_muts:\n",
    "\t\t\t\txobs[[\"missense\", \"nonsense\", \"coding-synon\", \"splice_site\"].index(mut[\"muttype\"])] += 1\n",
    "\t\texcept:\n",
    "\t\t\tpass\n",
    "\n",
    "\t\t###\tCompute neutral expectation (up to a constant factor). (synonymous, missense, nonsense)\n",
    "\t\texpect_x = [0. for t in range(4)]\n",
    "\t\tcod_cnt=0\n",
    "\t\tfor codon in codons_gene:\n",
    "\t\t\t# iterate over 3 context for each codon\n",
    "\t\t\tfor i in range(3):\n",
    "\t\t\t\tif context_mode:  #\tPENTANUCLEOTIDES\n",
    "\t\t\t\t\tif i==0:\n",
    "\t\t\t\t\t\tif cod_cnt==0:\n",
    "\t\t\t\t\t\t\tcontinue\n",
    "\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\tcur_pent = triplets[codons_gene[cod_cnt-1][2]]+triplets[codon[2]][:2]\n",
    "\t\t\t\t\telif i==1:\n",
    "\t\t\t\t\t\tcur_pent = triplets[codon[0]]+triplets[codon[2]][1:]\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\tif cod_cnt==len(codons_gene)-1:\n",
    "\t\t\t\t\t\t\tcontinue\n",
    "\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\tcur_pent = triplets[codon[1]]+triplets[codons_gene[cod_cnt+1][0]][1:]\n",
    "\t\t\t\t\n",
    "\t\t\t\tcodon_ID = codon[1]\n",
    "\t\t\t\tfor k in range(4):\n",
    "\t\t\t\t\tif context_mode==0:\t\t\t\t#\tTRINUCLEOTIDES\n",
    "\t\t\t\t\t\texpect_x[muttype_index(codon_ID)[i][k]] += neutral_matrix[codon[i]][k]\n",
    "\t\t\t\t\telse:\t\t\t\t\t\t\t#\tPENTANUCLEOTIDES\n",
    "\t\t\t\t\t\texpect_x[muttype_index(codon_ID)[i][k]] += neutral_matrix[quintuplets.index(cur_pent)][k]\n",
    "\t\t\tcod_cnt += 1\n",
    "\n",
    "\t\texp_obs_per_gene.append([cur_gene, expect_x[0], expect_x[1], expect_x[2], xobs[0], xobs[1], xobs[2], gene_len, xobs[3]])\n",
    "\t\t\n",
    "\treturn exp_obs_per_gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimize_neg_ln_L(p_start, function, mks_array, aux, bound_array, n_param):\n",
    "\tif n_param==2:\n",
    "\t\tp0, p1 = p_start\n",
    "\t\tres = minimize(function, (p0, p1), args=(mks_array, aux), method = 'L-BFGS-B', bounds = bound_array, options={'disp': None, 'gtol': 1e-12, 'eps': 1e-5, 'maxiter': 15000, 'ftol': 1e-12})\n",
    "\t\treturn [res.x[0], res.x[1], res.fun]\n",
    "\telif n_param==4:\n",
    "\t\tp0, p1, p2, p3 = p_start\n",
    "\t\tres = minimize(function, (p0, p1, p2, p3), args=(mks_array, aux), method = 'L-BFGS-B', bounds = bound_array, options={'disp': None, 'gtol': 1e-12, 'eps': 1e-5, 'maxiter': 15000, 'ftol': 1e-12})\n",
    "\t\treturn [res.x[0], res.x[1], res.x[2], res.x[3], res.fun]\n",
    "\telif n_param==5:\n",
    "\t\tp0, p1, p2, p3, p4 = p_start\n",
    "\t\tres = minimize(function, (p0, p1, p2, p3, p4), args=(mks_array, aux), method = 'L-BFGS-B', bounds = bound_array, options={'disp': None, 'gtol': 1e-12, 'eps': 1e-5, 'maxiter': 15000, 'ftol': 1e-12})\n",
    "\t\treturn [res.x[0], res.x[1], res.x[2], res.x[3], res.x[4], res.fun]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def neg_ln_L(p, genes, aux):\n",
    "\tmodC = aux\n",
    "\tif [1,2].count(modC):\n",
    "\t\ta,b = p\n",
    "\t\tif a<0:\n",
    "\t\t\ta=1e-6\n",
    "\t\tif b<0:\n",
    "\t\t\tb=1e-6\n",
    "\telif [3,4].count(modC):\n",
    "\t\ta,b,t,w = p\n",
    "\t\tif a<0:\n",
    "\t\t\ta=1e-6\n",
    "\t\tif b<0:\n",
    "\t\t\tb=1e-6\n",
    "\t\tif w<0:\n",
    "\t\t\tw=1e-6\n",
    "\t\tif t<0:\n",
    "\t\t\tt=1e-6\n",
    "\telif [5,6].count(modC):\n",
    "\t\ta,b,g,d,w = p\n",
    "\t\tif a<0:\n",
    "\t\t\ta=1e-6\n",
    "\t\tif b<0:\n",
    "\t\t\tb=1e-6\n",
    "\t\tif g<0:\n",
    "\t\t\tg=1e-6\n",
    "\t\tif d<0:\n",
    "\t\t\td=1e-6\n",
    "\t\tif w<0:\n",
    "\t\t\tw=1e-6\n",
    "\n",
    "\tgenes_by_sobs = [[ka, len(list(gr))] for ka, gr in it.groupby(sorted(genes, key=lambda arg: int(arg[\"obs\"][2])), key=lambda arg: int(arg[\"obs\"][2]))]\n",
    "\n",
    "\tsumme=0.\n",
    "\tif modC==1:\n",
    "\t\tfor sval in genes_by_sobs:\n",
    "\t\t\ts = sval[0]\n",
    "\t\t\t#*************** lambda ~ Gamma:\n",
    "\t\t\tsumme += sval[1] * (s*np.log(b) + (-s-a)*np.log(1 + b) + sp.gammaln(s + a) - sp.gammaln(s+1) - sp.gammaln(a))\n",
    "\n",
    "\telif modC==2:\n",
    "\t\tfor sval in genes_by_sobs:\n",
    "\t\t\ts = sval[0]\n",
    "\t\t\t#*************** lambda ~ IG:\n",
    "\t\t\tif s>25:\n",
    "\t\t\t\tsumme += sval[1] * (math.log(2.) + ((s + a)/2.)*math.log(b) + float(mp.log(mp.besselk(-s + a, 2*math.sqrt(b)).real)) - sp.gammaln(s+1) - sp.gammaln(a))\n",
    "\t\t\telse:\n",
    "\t\t\t\ttry:\n",
    "\t\t\t\t\tsumme += sval[1] * (math.log(2.) + ((s + a)/2.)*math.log(b) + math.log(sp.kv(-s + a, 2*math.sqrt(b))) - sp.gammaln(s+1) - sp.gammaln(a))\n",
    "\t\t\t\texcept:\n",
    "\t\t\t\t\tsumme += sval[1] * (math.log(2.) + ((s + a)/2.)*math.log(b) + float(mp.log(mp.besselk(-s + a, 2*math.sqrt(b)).real)) - sp.gammaln(s+1) - sp.gammaln(a))\n",
    "\t\t\t\t\t\n",
    "\telif modC==3:\n",
    "\t\tfor sval in genes_by_sobs:\n",
    "\t\t\ts = sval[0]\n",
    "\t\t\t#*************** lambda ~ w * Exp + (1-w) * Gamma:\n",
    "\t\t\tsumme += sval[1] * (np.log( math.exp( math.log(w * t) + (-1 - s)*math.log(1 + t) ) + math.exp( math.log(1.-w) + s*math.log(b) + (-s-a)*math.log(1 + b) + sp.gammaln(s + a) - sp.gammaln(s+1) - sp.gammaln(a) ) ))\n",
    "\telif modC==4:\n",
    "\t\tfor sval in genes_by_sobs:\n",
    "\t\t\ts = sval[0]\n",
    "\t\t\t#*************** lambda ~ w * Exp + (1-w) * InvGamma:\n",
    "\t\t\tif s>25:\n",
    "\t\t\t\tsumme += sval[1] * (np.log( math.exp( math.log(w * t) + (-1 - s)*math.log(1 + t) ) + math.exp( math.log(1.-w) + math.log(2.) + ((s + a)/2.)*math.log(b) + float(mp.log(mp.besselk(-s + a, 2*math.sqrt(b)).real)) - sp.gammaln(s + 1) - sp.gammaln(a) ) ))\n",
    "\t\t\telse:\n",
    "\t\t\t\ttry:\n",
    "\t\t\t\t\tsumme += sval[1] * (np.log( math.exp( math.log(w * t) + (-1 - s)*math.log(1 + t) ) + math.exp( math.log(1.-w) + math.log(2.) + ((s + a)/2.)*math.log(b) + math.log(sp.kv(-s + a, 2*math.sqrt(b))) - sp.gammaln(s + 1) - sp.gammaln(a) ) ))\n",
    "\t\t\t\texcept:\n",
    "\t\t\t\t\tsumme += sval[1] * (np.log( math.exp( math.log(w * t) + (-1 - s)*math.log(1 + t) ) + math.exp( math.log(1.-w) + math.log(2.) + ((s + a)/2.)*math.log(b) + float(mp.log(mp.besselk(-s + a, 2*math.sqrt(b)).real)) - sp.gammaln(s + 1) - sp.gammaln(a) ) ))\n",
    "\n",
    "\telif modC==5:\n",
    "\t\tfor sval in genes_by_sobs:\n",
    "\t\t\ts = sval[0]\n",
    "\t\t\t#*************** lambda ~ w * Gamma + (1-w) * Gamma:\n",
    "\t\t\tsumme += sval[1] * (np.log( np.exp( np.log(w) + s*np.log(b) + (-s-a)*np.log(1 + b) + sp.gammaln(s + a) - sp.gammaln(s+1) - sp.gammaln(a) ) + np.exp( np.log(1.-w) + s*np.log(d) + (-s-g)*np.log(1 + d) + sp.gammaln(s + g) - sp.gammaln(s+1) - sp.gammaln(g) ) ))\n",
    "\telif modC==6:\n",
    "\t\tfor sval in genes_by_sobs:\n",
    "\t\t\ts = sval[0]\n",
    "\t\t\t#*************** lambda ~ w * Gamma + (1-w) * InvGamma:\n",
    "\t\t\tif s>25:\n",
    "\t\t\t\tsumme += sval[1] * (np.log( (w * math.exp(s*math.log(b) + (-s-a)*math.log(1 + b) + sp.gammaln(s + a) - sp.gammaln(s+1) - sp.gammaln(a))) + ((1.-w) * math.exp( math.log(2.) + ((s + g)/2.)*math.log(d) + float(mp.log(mp.besselk(-s + g, 2*math.sqrt(d)).real)) - sp.gammaln(s + 1) - sp.gammaln(g) ) )))\n",
    "\t\t\telse:\n",
    "\t\t\t\ttry:\n",
    "\t\t\t\t\tsumme += sval[1] * (np.log( (w * b**s * (1 + b)**(-s-a) * math.exp(sp.gammaln(s + a) - sp.gammaln(s+1) - sp.gammaln(a))) + ((1.-w) * math.exp( math.log(2.) + ((s + g)/2.)*math.log(d) + math.log(sp.kv(-s + g, 2*math.sqrt(d))) - sp.gammaln(s + 1) - sp.gammaln(g) ) )))\n",
    "\t\t\t\texcept:\n",
    "\t\t\t\t\tsumme += sval[1] * (np.log( (w * math.exp(s*math.log(b) + (-s-a)*math.log(1 + b) + sp.gammaln(s + a) - sp.gammaln(s+1) - sp.gammaln(a))) + ((1.-w) * math.exp( math.log(2.) + ((s + g)/2.)*math.log(d) + float(mp.log(mp.besselk(-s + g, 2*math.sqrt(d)).real)) - sp.gammaln(s + 1) - sp.gammaln(g) ) )))\n",
    "\n",
    "\tlastres = -summe\n",
    "\tif lastres>1e8:\n",
    "\t\treturn 0\n",
    "\treturn -summe\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Jinxin\n",
    "#### BCC signature\n",
    "\n",
    "# import pandas as pd\n",
    "\n",
    "# def signatures_maf_data(infile):\n",
    "#     data_df = pd.read_csv(infile, sep='\\t', header=0)\n",
    "#     # data_gb = data_df.groupby([\"mut_nuc\", \"context\"]).size()\n",
    "#     data_gb = data_df.groupby([\"mut_nuc\", \"context\"])\n",
    "#     signatures = np.zeros(96)\n",
    "#     colnames   = []\n",
    "#     # for s in data_gb:\n",
    "#     #     print(s)\n",
    "#     #     break\n",
    "#     print(data_gb.get_group(('T',4)))\n",
    "    \n",
    "    # triplets.index(triplets_user[int(field[3])])\n",
    "    \n",
    "# signatures_maf_data(\"BCC_fullset.txt\")\n",
    "# data_df = pd.read_csv(\"BCC_fullset.txt\", sep='\\t', header=0)\n",
    "# set(data_df.mut_eff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\tmuttypes = [\"missense\", \"nonsense\", \"coding-synon\", \"intron\", \"utr-3\", \"utr-5\", \"IGR\"]\n",
    "triplets      = [\"AAA\", \"AAC\", \"AAG\", \"AAT\", \"CAA\", \"CAC\", \"CAG\", \"CAT\", \"GAA\", \"GAC\", \"GAG\", \"GAT\", \"TAA\", \"TAC\", \"TAG\", \"TAT\", \n",
    "                 \"ACA\", \"ACC\", \"ACG\", \"ACT\", \"CCA\", \"CCC\", \"CCG\", \"CCT\", \"GCA\", \"GCC\", \"GCG\", \"GCT\", \"TCA\", \"TCC\", \"TCG\", \"TCT\", \n",
    "                 \"AGA\", \"AGC\", \"AGG\", \"AGT\", \"CGA\", \"CGC\", \"CGG\", \"CGT\", \"GGA\", \"GGC\", \"GGG\", \"GGT\", \"TGA\", \"TGC\", \"TGG\", \"TGT\", \n",
    "                 \"ATA\", \"ATC\", \"ATG\", \"ATT\", \"CTA\", \"CTC\", \"CTG\", \"CTT\", \"GTA\", \"GTC\", \"GTG\", \"GTT\", \"TTA\", \"TTC\", \"TTG\", \"TTT\"]\n",
    "\n",
    "triplets_user = [\"AAA\", \"AAC\", \"AAG\", \"AAT\", \"ACA\", \"ACC\", \"ACG\", \"ACT\", \"AGA\", \"AGC\", \"AGG\", \"AGT\", \"ATA\", \"ATC\", \"ATG\", \"ATT\", \n",
    "                 \"CAA\", \"CAC\", \"CAG\", \"CAT\", \"CCA\", \"CCC\", \"CCG\", \"CCT\", \"CGA\", \"CGC\", \"CGG\", \"CGT\", \"CTA\", \"CTC\", \"CTG\", \"CTT\", \n",
    "                 \"GAA\", \"GAC\", \"GAG\", \"GAT\", \"GCA\", \"GCC\", \"GCG\", \"GCT\", \"GGA\", \"GGC\", \"GGG\", \"GGT\", \"GTA\", \"GTC\", \"GTG\", \"GTT\", \n",
    "                 \"TAA\", \"TAC\", \"TAG\", \"TAT\", \"TCA\", \"TCC\", \"TCG\", \"TCT\", \"TGA\", \"TGC\", \"TGG\", \"TGT\", \"TTA\", \"TTC\", \"TTG\", \"TTT\"]\n",
    "\n",
    "bases = [\"A\", \"C\", \"G\", \"T\", \"N\"]\n",
    "\n",
    "mod_choice_short\t= [\"\", \"G\", \"IG\", \"EmixG\", \"EmixIG\", \"GmixG\", \"GmixIG\"]\t\t\t# model choice\n",
    "\n",
    "mod_choice\t= [\"\", \"Gamma(a,b): [a,b] =\", \"InverseGamma(a,b): [a,b] =\", \"w Exp(t) + (1-w) Gamma(a,b): [a,b,t,w] =\", \"w Exp(t) + (1-w) InverseGamma(a,b): [a,b,t,w] =\", \"w Gamma(a,b) + (1-w) Gamma(g,d): [a,b,g,d,w] =\", \"w Gamma(a,b) + (1-w) InverseGamma(g,d): [a,b,g,d,w] =\"]\t\t# model choice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\tsomatic mutation data input file\n",
    "infile\t\t= \"resources/BCC_fullset_ss.txt\"\n",
    "# infile      = \"BCC_2k_subset_ss.txt\"\n",
    "\n",
    "#\tpath to auxiliary input files folder\n",
    "filepath\t= \"Auxiliary\"\n",
    "\n",
    "#\t0 = trinucleotides, 1 = pentanucleotides\n",
    "c_mode\t\t= 0\n",
    "\n",
    "#\tmodel choice: 0=all, 1=G, 2=IG, 3=EmixG, 4=EmixIG, 5=GmixG, 6=GmixIG\n",
    "mod_C\t\t= 0\n",
    "\n",
    "#\tname for the output file containing the q values\n",
    "outname\t\t= \"BCCfullset_jx\"\n",
    "# outname\t\t= \"BCC_2k_subset_ss_jx\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "rep_no = 30\t\t#\tNo. of independent runs to estimate model parameters (maximizing log-likelihood)\n",
    "\n",
    "quintuplets\t\t\t= import_quintuplets(\"%s/quintuplets.txt\" %filepath)\n",
    "cancer_genes\t\t= import_special_genes(\"%s/COSMIC_genes_v80.txt\" %filepath)\n",
    "essential_genes\t\t= import_special_genes(\"%s/Wang_cell_essential_genes.txt\" %filepath)\n",
    "zero_genes\t\t\t= import_special_genes(\"%s/zero_genes.txt\" %filepath)\n",
    "\n",
    "if c_mode==0:\n",
    "\tnuc_context_occs = import_context_freqs(\"%s/trinucleotide_occurrences_exons.txt\" %filepath)\n",
    "else:\n",
    "\tnuc_context_occs = import_context_freqs(\"%s/pentanucleotide_occurrences_exons.txt\" %filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Importing codons...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "#\t(1)\tImport mutation annotation file (maf) including header line, \"context\" is 0-based.\n",
    "\n",
    "#\tFormat: [\"gene\", \"muttype\", \"mutbase\", \"context\"]\n",
    "mutations = import_maf_data(infile, c_mode)\n",
    "lengths = [[k,len(list(g))] for k, g in it.groupby(sorted(mutations, key=lambda arg: arg[\"muttype\"]), key=lambda arg: arg[\"muttype\"])]\n",
    "\n",
    "#\t(2)\tImport trinucleotides, codons and splice sites for each gene.\n",
    "codons_by_gene   = import_codons_by_gene(\"%s/codons_by_gene.txt.gz\" %filepath)\n",
    "# sp_sites_by_gene = import_splice_sites_by_gene(\"%s/splice_sites_by_gene.txt\"%filepath)\n",
    "\n",
    "#\t(3)\tBuild array with neutral background mutations for construction of neutral mutation matrix.\n",
    "neutral_mutations = [mut for mut in mutations if [\"missense\", \"nonsense\", \"coding-synon\", \"utr-3\", \"utr-5\", \"splice_site\"].count(mut[\"muttype\"]) and cancer_genes.count(mut[\"gene\"])==0 and essential_genes.count(mut[\"gene\"])==0]\n",
    "\n",
    "neutral_muts_by_context = [[mut[\"context\"], bases.index(mut[\"mutbase\"])] for mut in neutral_mutations]\n",
    "\n",
    "print(\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "neutral_matrix: \n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "0.000000\t0.036931\t0.071686\t0.090489\t\n",
      "0.000000\t0.032326\t0.361193\t0.028948\t\n",
      "0.000000\t0.036391\t0.142220\t0.044119\t\n",
      "0.000000\t0.025165\t0.084392\t0.113807\t\n",
      "0.000000\t0.032903\t0.086750\t0.031400\t\n",
      "0.000000\t0.056303\t0.069859\t0.030411\t\n",
      "0.000000\t0.034903\t0.069658\t0.029152\t\n",
      "0.000000\t0.041052\t0.116269\t0.048699\t\n",
      "0.000000\t0.028510\t0.058046\t0.038441\t\n",
      "0.000000\t0.016362\t0.048856\t0.025580\t\n",
      "0.000000\t0.021280\t0.047803\t0.034302\t\n",
      "0.000000\t0.018417\t0.052652\t0.036834\t\n",
      "0.000000\t0.027683\t0.103503\t0.067868\t\n",
      "0.000000\t0.033361\t0.076228\t0.049384\t\n",
      "0.000000\t0.020522\t0.106968\t0.045886\t\n",
      "0.000000\t0.031723\t0.107517\t0.096689\t\n",
      "0.071050\t0.000000\t0.023806\t0.076217\t\n",
      "0.046734\t0.000000\t0.023471\t0.798177\t\n",
      "0.061493\t0.000000\t0.072084\t0.727322\t\n",
      "0.034205\t0.000000\t0.026664\t0.317388\t\n",
      "0.068970\t0.000000\t0.018111\t1.414074\t\n",
      "0.044157\t0.000000\t0.028895\t3.400464\t\n",
      "0.084671\t0.000000\t0.096810\t2.150561\t\n",
      "0.072217\t0.000000\t0.028924\t2.113260\t\n",
      "0.051771\t0.000000\t0.018190\t0.066727\t\n",
      "0.023021\t0.000000\t0.023909\t0.590912\t\n",
      "0.060804\t0.000000\t0.063847\t0.469414\t\n",
      "0.021494\t0.000000\t0.020003\t0.343426\t\n",
      "0.066878\t0.000000\t0.030073\t1.905736\t\n",
      "0.081532\t0.000000\t0.038696\t6.083732\t\n",
      "0.077618\t0.000000\t0.087014\t7.830063\t\n",
      "0.054145\t0.000000\t0.047880\t2.098435\t\n",
      "2.098435\t0.047880\t0.000000\t0.054145\t\n",
      "0.343426\t0.020003\t0.000000\t0.021494\t\n",
      "2.113260\t0.028924\t0.000000\t0.072217\t\n",
      "0.317388\t0.026664\t0.000000\t0.034205\t\n",
      "7.830063\t0.087014\t0.000000\t0.077618\t\n",
      "0.469414\t0.063847\t0.000000\t0.060804\t\n",
      "2.150561\t0.096810\t0.000000\t0.084671\t\n",
      "0.727322\t0.072084\t0.000000\t0.061493\t\n",
      "6.083732\t0.038696\t0.000000\t0.081532\t\n",
      "0.590912\t0.023909\t0.000000\t0.023021\t\n",
      "3.400464\t0.028895\t0.000000\t0.044157\t\n",
      "0.798177\t0.023471\t0.000000\t0.046734\t\n",
      "1.905736\t0.030073\t0.000000\t0.066878\t\n",
      "0.066727\t0.018190\t0.000000\t0.051771\t\n",
      "1.414074\t0.018111\t0.000000\t0.068970\t\n",
      "0.076217\t0.023806\t0.000000\t0.071050\t\n",
      "0.096689\t0.107517\t0.031723\t0.000000\t\n",
      "0.036834\t0.052652\t0.018417\t0.000000\t\n",
      "0.048699\t0.116269\t0.041052\t0.000000\t\n",
      "0.113807\t0.084392\t0.025165\t0.000000\t\n",
      "0.045886\t0.106968\t0.020522\t0.000000\t\n",
      "0.034302\t0.047803\t0.021280\t0.000000\t\n",
      "0.029152\t0.069658\t0.034903\t0.000000\t\n",
      "0.044119\t0.142220\t0.036391\t0.000000\t\n",
      "0.049384\t0.076228\t0.033361\t0.000000\t\n",
      "0.025580\t0.048856\t0.016362\t0.000000\t\n",
      "0.030411\t0.069859\t0.056303\t0.000000\t\n",
      "0.028948\t0.361193\t0.032326\t0.000000\t\n",
      "0.067868\t0.103503\t0.027683\t0.000000\t\n",
      "0.038441\t0.058046\t0.028510\t0.000000\t\n",
      "0.031400\t0.086750\t0.032903\t0.000000\t\n",
      "0.090489\t0.071686\t0.036931\t0.000000\t\n"
     ]
    }
   ],
   "source": [
    "#\t(4)\tDerive expected and observed mutation counts for all three categories per gene.\n",
    "res = export_expected_observed_mks_per_gene(codons_by_gene, mutations, neutral_muts_by_context, nuc_context_occs, c_mode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "mks_type=[]\n",
    "\n",
    "for gene in res:\n",
    "\tmks_type.append({\"gene\": gene[0], \"exp\": [float(gene[1]), float(gene[2]), float(gene[3])], \"obs\": [int(gene[4]), int(gene[5]), int(gene[6]), int(gene[8])], \"len\": int(gene[7])})\n",
    "\n",
    "mks_type = sorted(mks_type, key=lambda arg: arg[\"gene\"])\n",
    "mks_type = [mks for mks in mks_type if (len(mks[\"gene\"])>2 and mks[\"gene\"][:2]==\"OR\" and ['0','1','2','3','4','5','6','7'].count(mks[\"gene\"][2]))==0]\n",
    "mks_type = [gene for gene in mks_type if zero_genes.count(gene[\"gene\"])==0]\n",
    "\n",
    "fout = open(\"Output/output_data_preparation_%s.txt\" %outname, \"w\")\n",
    "\n",
    "# Output format: [gene, lm, lk, ls, mobs, kobs, sobs, Lgene, cobs]\n",
    "for gene in mks_type:\n",
    "\tfout.write(\"%s\\t%f\\t%f\\t%i\\t%i\\t%i\\t%i\\t%i\\t%i\\n\" %(gene[\"gene\"], gene[\"exp\"][0], gene[\"exp\"][1], gene[\"exp\"][2], gene[\"obs\"][0], gene[\"obs\"][1], gene[\"obs\"][2], gene[\"len\"], gene[\"obs\"][3]))\n",
    "fout.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting model 1...\n",
      "100% done.\r"
     ]
    }
   ],
   "source": [
    "if mod_C==1 or mod_C==0:\n",
    "\tsys.stdout.write(\"Fitting model 1...\\n\")\n",
    "\tlow_b\t= [1e-5*random.uniform(1.,3.) for i in range(2)]\n",
    "\tup_b\t= [50.*random.uniform(1.,2.) for i in range(2)]\n",
    "\tcur_min_res=[0,0,1e20]\n",
    "\tfor rep in range(rep_no):\n",
    "\t\tsys.stdout.write(\"%.f%% done.\\r\" %(100.*(rep+1)/rep_no))\n",
    "\t\tp_res = minimize_neg_ln_L([random.uniform(0.02,10.), random.uniform(0.02,10.)], neg_ln_L, mks_type, 1, [(low_b[0],up_b[0]), (low_b[1],up_b[1])], 2)\n",
    "\t\tif p_res[2]>0 and p_res[2]<cur_min_res[2]:\n",
    "\t\t\tcur_min_res = p_res[:]\n",
    "\tif cur_min_res[2]==1e20:\n",
    "\t\tsys.stderr.write(\"Could not find a converging solution for model 1.\\n\")\n",
    "\tfout = open(\"Output/param_estimates_%s_1.txt\" %outname, \"w\")\n",
    "\tfout.write(\"%e, %e, %f, %i\\n\" %(cur_min_res[0], cur_min_res[1], cur_min_res[2], 1))\n",
    "\tfout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting model 2...\n",
      "100% done.\r"
     ]
    }
   ],
   "source": [
    "if mod_C==2 or mod_C==0:\n",
    "\tsys.stdout.write(\"Fitting model 2...\\n\")\n",
    "\tlow_b\t= [1e-5*random.uniform(1.,3.) for i in range(2)]\n",
    "\tup_b\t= [50.*random.uniform(1.,2.) for i in range(2)]\n",
    "\tcur_min_res=[0,0,1e20]\n",
    "\tfor rep in range(rep_no):\n",
    "\t\tsys.stdout.write(\"%.f%% done.\\r\" %(100.*(rep+1)/rep_no))\n",
    "\t\tp_res = minimize_neg_ln_L([random.uniform(0.02,10.), random.uniform(0.02,10.)], neg_ln_L, mks_type, 2, [(low_b[0],up_b[0]), (low_b[1],up_b[1])], 2)\n",
    "\t\tif p_res[2]>0 and p_res[2]<cur_min_res[2]:\n",
    "\t\t\tcur_min_res = p_res[:]\n",
    "\tif cur_min_res[2]==1e20:\n",
    "\t\tsys.stderr.write(\"Could not find a converging solution for model 2.\\n\")\n",
    "\tfout = open(\"Output/param_estimates_%s_2.txt\" %outname, \"w\")\n",
    "\tfout.write(\"%e, %e, %f, %i\\n\" %(cur_min_res[0], cur_min_res[1], cur_min_res[2], 2))\n",
    "\tfout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting model 3...\n",
      "27% done.\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/beegfs/scratch/j_wang/jupyter-3700932/ipykernel_42969/2281193467.py:57: RuntimeWarning: divide by zero encountered in log\n",
      "  summe += sval[1] * (np.log( math.exp( math.log(w * t) + (-1 - s)*math.log(1 + t) ) + math.exp( math.log(1.-w) + s*math.log(b) + (-s-a)*math.log(1 + b) + sp.gammaln(s + a) - sp.gammaln(s+1) - sp.gammaln(a) ) ))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100% done.\r"
     ]
    }
   ],
   "source": [
    "if mod_C==3 or mod_C==0:\n",
    "\tsys.stdout.write(\"Fitting model 3...\\n\")\n",
    "\tlow_b\t= [1e-5*random.uniform(1.,3.) for i in range(4)]\n",
    "\tup_b\t= [50.*random.uniform(1.,2.) for i in range(4)]\n",
    "\tcur_min_res=[0,0,0,0,1e20]\n",
    "\tfor rep in range(rep_no):\n",
    "\t\tsys.stdout.write(\"%.f%% done.\\r\" %(100.*(rep+1)/rep_no))\n",
    "\t\tp_res = minimize_neg_ln_L([random.uniform(0.02,10.), random.uniform(0.02,10.), random.uniform(0.02,10.), random.uniform(2e-5,0.95)], neg_ln_L, mks_type, 3, [(low_b[0],up_b[0]), (low_b[1],up_b[1]), (low_b[2],up_b[2]), (low_b[3],0.9999)], 4)\n",
    "\t\tif p_res[4]>0 and p_res[4]<cur_min_res[4]:\n",
    "\t\t\tcur_min_res = p_res[:]\n",
    "\tif cur_min_res[4]==1e20:\n",
    "\t\tsys.stderr.write(\"Could not find a converging solution for model 3.\\n\")\n",
    "\tfout = open(\"Output/param_estimates_%s_3.txt\" %outname, \"w\")\n",
    "\tfout.write(\"%e, %e, %e, %e, %f, %i\\n\" %(cur_min_res[0], cur_min_res[1], cur_min_res[2], cur_min_res[3], cur_min_res[4], 3))\n",
    "\tfout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting model 4...\n",
      "100% done.\r"
     ]
    }
   ],
   "source": [
    "if mod_C==4 or mod_C==0:\n",
    "\tsys.stdout.write(\"Fitting model 4...\\n\")\n",
    "\tlow_b\t= [1e-5*random.uniform(1.,3.) for i in range(4)]\n",
    "\tup_b\t= [50.*random.uniform(1.,2.) for i in range(4)]\n",
    "\tcur_min_res=[0,0,0,0,1e20]\n",
    "\tfor rep in range(rep_no):\n",
    "\t\tsys.stdout.write(\"%.f%% done.\\r\" %(100.*(rep+1)/rep_no))\n",
    "\t\tp_res = minimize_neg_ln_L([random.uniform(0.02,10.), random.uniform(0.02,10.), random.uniform(0.02,10.), random.uniform(2e-5,0.95)], neg_ln_L, mks_type, 4, [(low_b[0],up_b[0]), (low_b[1],up_b[1]), (low_b[2],up_b[2]), (low_b[3],0.9999)], 4)\n",
    "\t\tif p_res[4]>0 and p_res[4]<cur_min_res[4]:\n",
    "\t\t\tcur_min_res = p_res[:]\n",
    "\tif cur_min_res[4]==1e20:\n",
    "\t\tsys.stderr.write(\"Could not find a converging solution for model 4.\\n\")\n",
    "\tfout = open(\"Output/param_estimates_%s_4.txt\" %outname, \"w\")\n",
    "\tfout.write(\"%e, %e, %e, %e, %f, %i\\n\" %(cur_min_res[0], cur_min_res[1], cur_min_res[2], cur_min_res[3], cur_min_res[4], 4))\n",
    "\tfout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting model 5...\n",
      "7% done.\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/beegfs/scratch/j_wang/jupyter-3700932/ipykernel_42969/2281193467.py:74: RuntimeWarning: divide by zero encountered in log\n",
      "  summe += sval[1] * (np.log( np.exp( np.log(w) + s*np.log(b) + (-s-a)*np.log(1 + b) + sp.gammaln(s + a) - sp.gammaln(s+1) - sp.gammaln(a) ) + np.exp( np.log(1.-w) + s*np.log(d) + (-s-g)*np.log(1 + d) + sp.gammaln(s + g) - sp.gammaln(s+1) - sp.gammaln(g) ) ))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100% done.\r"
     ]
    }
   ],
   "source": [
    "if mod_C==5 or mod_C==0:\n",
    "\tsys.stdout.write(\"Fitting model 5...\\n\")\n",
    "\tlow_b\t= [1e-5*random.uniform(1.,3.) for i in range(5)]\n",
    "\tup_b\t= [50.*random.uniform(1.,2.) for i in range(5)]\n",
    "\tcur_min_res=[0,0,0,0,0,1e20]\n",
    "\tfor rep in range(int(rep_no)):\n",
    "\t\tsys.stdout.write(\"%.f%% done.\\r\" %(100.*(rep+1)/rep_no))\n",
    "\t\tp_res = minimize_neg_ln_L([random.uniform(0.02,10.), random.uniform(0.02,5.), random.uniform(0.02,10.), random.uniform(0.02,10.), random.uniform(2e-5,0.95)], neg_ln_L, mks_type, 5, [(low_b[0],up_b[0]), (low_b[1],up_b[1]), (low_b[2],up_b[2]), (low_b[3],up_b[3]), (low_b[4],0.9999)], 5)\n",
    "\t\tif p_res[5]>0 and p_res[5]<cur_min_res[5]:\n",
    "\t\t\tcur_min_res = p_res[:]\n",
    "\tif cur_min_res[5]==1e20:\n",
    "\t\tsys.stderr.write(\"Could not find a converging solution for model 5.\\n\")\n",
    "\tfout = open(\"Output/param_estimates_%s_5.txt\" %outname, \"w\")\n",
    "\tfout.write(\"%e, %e, %e, %e, %e, %f, %i\\n\" %(cur_min_res[0], cur_min_res[1], cur_min_res[2], cur_min_res[3], cur_min_res[4], cur_min_res[5], 5))\n",
    "\tfout.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting model 6...\n",
      "5% done.\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/beegfs/scratch/j_wang/jupyter-3700932/ipykernel_42969/2281193467.py:80: RuntimeWarning: divide by zero encountered in log\n",
      "  summe += sval[1] * (np.log( (w * math.exp(s*math.log(b) + (-s-a)*math.log(1 + b) + sp.gammaln(s + a) - sp.gammaln(s+1) - sp.gammaln(a))) + ((1.-w) * math.exp( math.log(2.) + ((s + g)/2.)*math.log(d) + float(mp.log(mp.besselk(-s + g, 2*math.sqrt(d)).real)) - sp.gammaln(s + 1) - sp.gammaln(g) ) )))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100% done.\r"
     ]
    }
   ],
   "source": [
    "if mod_C==6 or mod_C==0:\n",
    "\tsys.stdout.write(\"Fitting model 6...\\n\")\n",
    "\tlow_b\t= [1e-5*random.uniform(1.,3.) for i in range(5)]\n",
    "\tup_b\t= [50.*random.uniform(1.,2.) for i in range(5)]\n",
    "\tcur_min_res=[0,0,0,0,0,1e20]\n",
    "\tfor rep in range(int(2*rep_no)):\n",
    "\t\tsys.stdout.write(\"%.f%% done.\\r\" %(100.*(rep+1)/(2*rep_no)))\n",
    "\t\tp_res = minimize_neg_ln_L([random.uniform(0.02,10.), random.uniform(0.02,5.), random.uniform(0.02,10.), random.uniform(0.02,10.), random.uniform(2e-5,0.95)], neg_ln_L, mks_type, 6, [(low_b[0],up_b[0]), (low_b[1],up_b[1]), (low_b[2],up_b[2]), (low_b[3],up_b[3]), (low_b[4],0.9999)], 5)\n",
    "\t\tif p_res[5]>0 and p_res[5]<cur_min_res[5]:\n",
    "\t\t\tcur_min_res = p_res[:]\n",
    "\tif cur_min_res[5]==1e20:\n",
    "\t\tsys.stderr.write(\"Could not find a converging solution for model 6.\\n\")\n",
    "\tfout = open(\"Output/param_estimates_%s_6.txt\" %outname, \"w\")\n",
    "\tfout.write(\"%e, %e, %e, %e, %e, %f, %i\\n\" %(cur_min_res[0], cur_min_res[1], cur_min_res[2], cur_min_res[3], cur_min_res[4], cur_min_res[5], 6))\n",
    "\tfout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
